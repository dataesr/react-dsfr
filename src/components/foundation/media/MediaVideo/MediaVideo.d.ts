import * as React from 'react';

export type MediaVideoRatio = '' | '4x3';

export type MediaVideoSize = 'sm' | 'md' | 'lg';

export interface MediaVideoProps {
  buttonLabel?: string;
  children: React.ReactNode;
  className?: string;
  legend?: string;
  onTranscriptionClick?: (...args: any[])=>any;
  ratio?: MediaVideoRatio;
  size?: MediaVideoSize;
}

declare const MediaVideo: React.FC<MediaVideoProps>;

export default MediaVideo;
