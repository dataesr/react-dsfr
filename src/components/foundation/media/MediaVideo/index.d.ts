import MediaVideo from "./MediaVideo"

export { MediaVideo }
