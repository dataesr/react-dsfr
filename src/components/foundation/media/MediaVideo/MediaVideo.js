import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import dataAttributes from '../../../../utils/data-attributes';

/**
 *
 * @visibleName MediaVideo
 */
const MediaVideo = ({
  buttonLabel,
  className,
  children,
  legend,
  onTranscriptionClick,
  ratio,
  size,
  ...remainingProps
}) => {
  const _className = classNames(
    'fr-content-media',
    `fr-content-media--${size}`,
    className,
  );
  const _classNameRatio = classNames(
    'fr-responsive-vid',
    { 'fr-responsive-vid--4x3': ratio },
  );

  return (
    <div
      {...dataAttributes.getAll(remainingProps)}
      className={_className}
    >
      <div className={_classNameRatio}>
        {children}
      </div>
      <div className="fr-content-media__caption">{legend}</div>
      {
        (buttonLabel) ? (
          <div className="fr-content-media__transcription">
            <button
              className="fr-btn"
              title={buttonLabel}
              type="button"
              onClick={onTranscriptionClick}
            >
              {buttonLabel}
            </button>
          </div>
        ) : null
      }
    </div>
  );
};

MediaVideo.defaultProps = {
  buttonLabel: '',
  className: '',
  legend: '',
  onTranscriptionClick: () => {},
  ratio: '',
  size: 'md',
};

MediaVideo.propTypes = {
  buttonLabel: PropTypes.string,
  children: PropTypes.node.isRequired,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  legend: PropTypes.string,
  onTranscriptionClick: PropTypes.func,
  ratio: PropTypes.oneOf(['', '4x3']),
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
};

export default MediaVideo;
