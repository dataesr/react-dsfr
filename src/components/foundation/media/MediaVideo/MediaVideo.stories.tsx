import { ComponentStory } from '@storybook/react';

import MediaVideo from './MediaVideo';

export default {
  title: 'MediaVideo',
  component: MediaVideo,
  argTypes: {
    size: {
      description: `Si vous souhaitez accentuer la visibilité d'une vidéo vous pouvez utiliser les différentes tailles à votre disposition
      - lg : dans cette taille, la largeur de la vidéo est supérieure à celle du contenu.
      - md : dans cette taille, la largeur de la vidéo correspond à celle du contenu.
      - sm : dans cette taille, la largeur de la vidéo est inférieure à celle du contenu.
      `,
    },
  },
};

const Template: ComponentStory<typeof MediaVideo> = (args) => (
  <MediaVideo {...args}>
    <iframe title="titre de l’iframe" className="fr-responsive-vid__player" src="https://www.youtube.com/embed/HyirpmPL43I" width="100%" height="100%" frameBorder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowFullScreen="" />
  </MediaVideo>
);

export const Video16x9 = Template.bind({});
Video16x9.args = {
  buttonLabel: 'Transcription',
  legend: 'legend',
  size: 'md',
};
Video16x9.storyName = 'Video 16x9';

export const Video4x3 = Template.bind({});
Video4x3.args = {
  buttonLabel: 'Transcription',
  legend: 'legend',
  size: 'md',
  ratio: '4x3',
};
Video4x3.storyName = 'Video 4x3';
