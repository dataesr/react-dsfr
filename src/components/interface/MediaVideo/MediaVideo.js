import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

/**
 *
 * @visibleName MediaVideo
 */
const MediaVideo = ({
  buttonLabel,
  className,
  children,
  legend,
  onTranscriptionClick,
  ratio,
  size,
}) => {
  const _className = classNames(
    'rf-content-media',
    `rf-content-media--${size}`,
    className,
  );
  const _classNameRatio = classNames(
    'rf-responsive-vid',
    { 'rf-responsive-vid--4x3': ratio },
  );

  return (
    <div className={_className} data-testid="video-test">
      <div className={_classNameRatio}>
        {children}
      </div>
      <div className="rf-content-media__caption">{legend}</div>
      {
        (buttonLabel) ? (
          <div className="rf-content-media__transcription">
            <button
              className="rf-btn"
              title={buttonLabel}
              type="button"
              onClick={onTranscriptionClick}
            >
              {buttonLabel}
            </button>
          </div>
        ) : null
      }
    </div>
  );
};

MediaVideo.defaultProps = {
  buttonLabel: '',
  className: '',
  legend: '',
  onTranscriptionClick: () => {},
  ratio: '',
  size: 'md',
};

MediaVideo.propTypes = {
  buttonLabel: PropTypes.string,
  children: PropTypes.node.isRequired,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  legend: PropTypes.string,
  onTranscriptionClick: PropTypes.func,
  ratio: PropTypes.oneOf(['', '4x3']),
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
};

export default MediaVideo;
