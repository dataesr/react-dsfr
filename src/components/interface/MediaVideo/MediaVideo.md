## Médias (intégrés au contenu) - Responsive medias
Les médias désignent vos contenus photos et vidéos. Lorsqu’ils sont intégré à une page de contenu, il est recommandé de suivre les règles décrites ci-dessous.

### Vidéo intégrée dans un contenu
Lorsqu’une vidéo est intégrée dans un contenu éditorial, elle est présentée par défaut en format 16:9 et peut intégrer une légende et un lien pour afficher la transcription. L’Iframe générée par l’hébergeur de la vidéo peut être différent du ratio par défaut de 16:9.

Il est possible d’opter pour un format 4:3 de la vidéo, à l’aide du modifier .rf-responsive-vid--4x3.

#### Ratio 16:9 :
```jsx
  <MediaVideo
    legend="Legend"
    buttonLabel="Button Label"
  >
    <iframe title="titre de l’iframe" class="rf-responsive-vid__player" src="https://www.youtube.com/embed/HyirpmPL43I" width="100%" height="100%" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="" />
  </MediaVideo>
```

#### Ratio 4:3 :
```jsx
  <MediaVideo
    legend="Legend"
    ratio="4x3"
    buttonLabel="Transcription"
  >
    <iframe title="titre de l’iframe" class="rf-responsive-vid__player" src="https://www.youtube.com/embed/HyirpmPL43I" width="100%" height="100%" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="" />
  </MediaVideo>
```

### Les trois tailles proposées pour les vidéos
Si vous souhaitez accentuer la visibilité d'une vidéo vous pouvez utiliser les différentes tailles à votre disposition
- LG : dans cette taille, la largeur de la vidéo est supérieure à celle du contenu.
- MD : dans cette taille, la largeur de la vidéo correspond à celle du contenu.
- SM : dans cette taille, la largeur de la vidéo est inférieure à celle du contenu.

#### Vidéo taille LG (125%)
La vidéo est plus grande que la zone de texte.
Exemple de vidéo LG format 16/9 :
```jsx
  <MediaVideo
    legend="Legend"
    size="lg"
    buttonLabel="Transcription"
  >
    <iframe title="titre de l’iframe" class="rf-responsive-vid__player" src="https://www.youtube.com/embed/HyirpmPL43I" width="100%" height="100%" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="" />
  </MediaVideo>
```
Exemple de vidéo LG format 4/3 :
```jsx
  <MediaVideo
    legend="Legend"
    size="lg"
    ratio="4X3"
    buttonLabel="Transcription"
  >
    <iframe title="titre de l’iframe" class="rf-responsive-vid__player" src="https://www.youtube.com/embed/HyirpmPL43I" width="100%" height="100%" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="" />
  </MediaVideo>
```

#### Vidéo taille SM (75%)
La largeur de la vidéo est plus petite que la zone de texte.
Exemple de vidéo SM format 16/9 :
```jsx
  <MediaVideo
    legend="Legend"
    size="sm"
    buttonLabel="Transcription"
  >
    <iframe title="titre de l’iframe" class="rf-responsive-vid__player" src="https://www.youtube.com/embed/HyirpmPL43I" width="100%" height="100%" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="" />
  </MediaVideo>
```
Exemple de vidéo SM format 4/3 :
```jsx
  <MediaVideo
    legend="Legend"
    size="sm"
    ratio="4X3"
    buttonLabel="Transcription"
  >
    <iframe title="titre de l’iframe" class="rf-responsive-vid__player" src="https://www.youtube.com/embed/HyirpmPL43I" width="100%" height="100%" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="" />
  </MediaVideo>
```
