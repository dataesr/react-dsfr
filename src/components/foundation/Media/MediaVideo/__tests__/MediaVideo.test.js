import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import MediaVideo from '../index';

describe('<MediaVideo />', () => {
  it('renders correctly', () => {
    const component = renderer
      .create(
        <MediaVideo
          legend="légende de l'image"
          size="sm"
          buttonLabel="Transcription"
        >
          <iframe
            title="titre de l’iframe"
            className="fr-responsive-vid__player"
            src="https://www.youtube.com/embed/HyirpmPL43I"
            width="100%"
            height="100%"
            frameBorder="0"
            allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
            allowFullScreen=""
          />
        </MediaVideo>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('in document', () => {
    render(
      <MediaVideo
        legend="légende de l'image"
        size="sm"
        buttonLabel="Transcription"
      >
        <iframe
          title="titre de l’iframe"
          className="fr-responsive-vid__player"
          src="https://www.youtube.com/embed/HyirpmPL43I"
          width="100%"
          height="100%"
          frameBorder="0"
          allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen=""
        />
      </MediaVideo>,
    );
    const search = screen.getByTestId('video');
    expect(search).toBeInTheDocument();
  });
});
