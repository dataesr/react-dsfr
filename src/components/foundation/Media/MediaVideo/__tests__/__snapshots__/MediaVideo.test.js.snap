// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MediaVideo /> renders correctly 1`] = `
<div
  className="fr-content-media fr-content-media--sm"
  data-testid="video"
>
  <div
    className="fr-responsive-vid"
  >
    <iframe
      allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
      allowFullScreen=""
      className="fr-responsive-vid__player"
      frameBorder="0"
      height="100%"
      src="https://www.youtube.com/embed/HyirpmPL43I"
      title="titre de l’iframe"
      width="100%"
    />
  </div>
  <div
    className="fr-content-media__caption"
  >
    légende de l'image
  </div>
  <div
    className="fr-content-media__transcription"
  >
    <button
      className="fr-btn"
      onClick={[Function]}
      title="Transcription"
      type="button"
    >
      Transcription
    </button>
  </div>
</div>
`;
